Mathlib/Data/Nat/Factorial/SuperFactorial.lean

```diff
@@ -79,6 +84,100 @@ theorem det_vandermonde_id_eq_superFactorial (n : ℕ) :
+    (Matrix.vandermonde v).det = (Matrix.of (fun i j => ((p j).eval (v i)))).det := by
+  rw [eval_matrixOfPolynomials_eq_vandermonde_mul_matrixOfPolynomials v p h_deg, Matrix.det_mul,
+      det_matrixOfPolynomials p h_deg h_monic, mul_one]
+
+theorem eval_matrixOfPolynomials_eq_mul_matrix_of_choose {n : ℕ} (v : Fin n → ℕ) :
+    (Matrix.of (fun (i j : Fin n) =>
+    ((fun n => descPochhammer ℤ n) j).eval ((fun k => (v k : ℤ)) i)) ).det =
+    (∏ i : Fin n,  Nat.factorial i) *
+    (Matrix.of (fun (i j : Fin n)  => ((Nat.choose ((v i)) (j : ℕ)) : ℤ))).det := by
+  convert Matrix.det_mul_row (fun (i : Fin n)  => ((Nat.factorial (i : ℕ)):ℤ)) _
+  · rw [Matrix.of_apply, descPochhammer_int_eq_descFactorial _ _]
+    congr
+    exact Nat.descFactorial_eq_factorial_mul_choose _ _
+  · rw [Nat.cast_prod]
+
+theorem superFactorial_dvd_vandermonde_det {n : ℕ} (v : Fin (n + 1) → ℤ) :
+    ↑(Nat.superFactorial n) ∣ (Matrix.vandermonde v).det := by
+  let m' := List.minimum ((univ : Finset (Fin (n + 1))).image v).toList
+  have : ((univ : Finset (Fin (n + 1))).image v).toList ≠ List.nil := by
+    simp only [ne_eq, toList_eq_nil, image_eq_empty]
+    refine Nonempty.ne_empty univ_nonempty
+  cases' Option.isSome_iff_exists.mp <| Option.ne_none_iff_isSome.mp <|
+      List.minimum_ne_top_of_ne_nil this with m hm
+  have h_min : ∀  l, l ∈ ((univ : Finset (Fin (n + 1))).image v).toList →  m' ≤ l := by
+     exact fun l a ↦ List.minimum_le_of_mem' a
+  let w' := fun i ↦ (v i - m).toNat
+  have h1 := eval_matrixOfPolynomials_eq_mul_matrix_of_choose w'
+  rw [← prod_range_factorial_succ, ← Fin.prod_univ_eq_prod_range]
+  have h2 := vandermonde_det_eq_eval_matrixOfPolynomials_det (fun i ↦ ↑(w' i))
+      (fun i => descPochhammer ℤ i)
+      (fun i => descPochhammer_natDegree ℤ i)
+      (fun i => monic_descPochhammer ℤ i)
+  rw [← h2] at h1
+  have : Matrix.det (Matrix.vandermonde fun i ↦ ↑(w' i)) = (Matrix.vandermonde v).det := by
+    rw [← Matrix.det_vandermonde_sub v m]
+    congr
+    unfold Matrix.vandermonde
+    ext i j
+    congr
+    simp only [sub_nonneg]
+    have : m ≤ v i  := by
+      have v_in : (v i) ∈ (image v univ) := by
+        unfold Finset.image
+        simp only [Multiset.mem_toFinset, Multiset.mem_map, mem_val, mem_univ, true_and,
+            exists_apply_eq_apply]
+      have  v_in' : ↑(v i) ∈ toList (image v univ) := mem_toList.mpr v_in
+      have h_min := h_min (v i) v_in'
+      simp only [hm] at h_min
+      rw [WithTop.some_eq_coe] at h_min
+      norm_num at h_min
+      exact h_min
+    exact Int.toNat_sub_of_le this
+  rw [this] at h1
+  norm_num at h1
+  use (Matrix.of (fun (i j : Fin (n + 1))  => ((Nat.choose ((v i  - m).toNat) (j : ℕ)) : ℤ))).det
+  convert h1
+  norm_cast
+  rw [Fin.prod_univ_eq_prod_range, ← Ico_zero_eq_range,
+      Finset.prod_eq_prod_Ico_succ_bot <| succ_pos n, ← Finset.prod_Ico_add' (fun x => x !) 0 n 1,
+      Ico_zero_eq_range, ← Fin.prod_univ_eq_prod_range, factorial_zero, one_mul]
```

Comment:
Using the following result:
```lean
@[simp]
theorem prod_range_succ_factorial : ∀ n : ℕ, ∏ x in range (n + 1), x ! = sf n
  | 0 => rfl
  | n + 1 => by
    rw [prod_range_succ, prod_range_succ_factorial n, mul_comm, superFactorial]
```
and the definition of `m` above, we can simplify this proof down to:
```suggestion
  have hw' : ∀ i, (w' i : ℤ) = v i - m := fun i ↦ Int.toNat_sub_of_le (inf'_le _ (mem_univ _))
  have h := vandermonde_det_eq_eval_matrixOfPolynomials_det (fun i ↦ ↑(w' i))
      (fun i => descPochhammer ℤ i)
      (fun i => descPochhammer_natDegree ℤ i)
      (fun i => monic_descPochhammer ℤ i)
  conv_lhs at h => simp only [hw', Matrix.det_vandermonde_sub]
  use (Matrix.of (fun (i j : Fin (n + 1)) => (Nat.choose (w' i) (j : ℕ) : ℤ))).det
  simp [h, eval_matrixOfPolynomials_eq_mul_matrix_of_choose w', Fin.prod_univ_eq_prod_range]
```